tests/chgres_cube/ftst_example.F90

```diff
@@ -0,0 +1,83 @@
+! Unit test for rh2spfh in grib2_utils to be used as an example for users
+! learning how to write unit tests. Users are prompted to add an additional
+! test for convert_omega
+! Larissa Reames OU/CIMMS/NOAA/NSSL/FRDD
+ 
+program ftst_example
+
+ use esmf
+ use model_grid, only : i_input, j_input 
+ use grib2_util, only : rh2spfh
+ 
+ implicit none
+
+ real(esmf_kind_r4), allocatable :: rh_spfh(:,:), &
+                                    rh_orig(:,:), &
+                                    spfh_returned(:,:), &
+                                    spfh_correct(:,:)
+ real(esmf_kind_r8),allocatable  :: t(:,:)
+ real(esmf_kind_r8) :: p
+ real,parameter          :: EPS = 1.0E-6
+
+ i_input = 2
+ j_input = 2
+ allocate(rh_spfh(i_input,j_input))
+ allocate(rh_orig(i_input,j_input))
+ allocate(spfh_returned(i_input,j_input))
+ allocate(spfh_correct(i_input,j_input))
+ allocate(t(i_input,j_input))
+
+ ! -------------------------------------------------------------------------
+ ! Set constants/values to be passed to the unit test. In this case it's a
+ ! set of single values, but it could be more complicated, like an 
+ ! n-dimensional array or ESMF objects.
+ ! -------------------------------------------------------------------------
+
+ rh_spfh(:,:) = 50.0  ! Relative humidity (%)
+ p = 100000.0    ! Pressure (Pa)
+ t(:,:) = 300.0       ! Temperature (K)
+ spfh_correct(:,:) = 10.978297E-3 ! Correct specific humidity value (kg/kg)
+
+ print*, "Starting Unit Testing rh2spfh."
+
+ !-------------------------------------------------------------------------
+ ! Execute testing below by running target function rh2spfh and providing
+ ! values set above
+ !-------------------------------------------------------------------------
+
+ rh_orig = rh_spfh          ! Save the original RH value for posterity
+ call rh2spfh(rh_spfh,p,t) 
+ spfh_returned = rh_spfh    ! Rename the returned value for clarity
+
+ !-------------------------------------------------------------------------
+ ! Check the returned value against what we know to be the correct answer. 
+ ! If the correct result is returned (within a certain small tolerance), 
+ ! then the test passes and the called routine is working as expected. If the 
+ ! incorrect value is passed back, the test fails and an error is returned.
+ !-------------------------------------------------------------------------
+
+ if ( any(abs(spfh_returned - spfh_correct) .lt. EPS)) then
+   print*, "RH2SPFH TEST PASSED. SUCCESS!"
+ else
+   print*, "RH2SPFH TEST FAILED."
+   print*, "TEST RETURNED VALUE OF ", spfh_returned
+   print*, "RETURNED VALUE EXPECT TO BE ", spfh_correct
+   stop 1
+ endif
```

Comment:
I am happy to see this sample test, and I am delighted to see any test going into the code base! So thanks for that.

As someone who has written many tens of thousands of lines of test code, let me suggest that a more compact idiom works just as well, with a lot less typing.

Here's some code from a test I wrote today that illustrates the point. Note that instead of 7 lines of code to check a value (as above) there is only one line per check. The stop code allows the developer to see exactly what test failed, without any print statements needed:

```
  print *, 'Testing getdrstemplate...'
  call getdrstemplate(0, nummap, map, needext, iret)
  if (iret .ne. 0 .or. nummap .ne. 5 .or. needext) stop 4
  if (map(1) .ne. 4 .or. map(2) .ne. -2 .or. map(3) .ne. -2 .or. &
       map(4) .ne. 1 .or. map(5) .ne. 1) stop 5

```

Note that instead of any print statements, every test is of the form:

if (something .ne. expected_value) stop 22

Where the number give to stop changes throughout the code. When a test fails, you get the stop code, which allows you to zero in on exactly the test that is failing.

Reply:
I commented out the majority of the the more wordy output, instructing the user that they may wish to use that sort of code to debug any test they create in the future. I did leave a single line of test output indicating the name of the failed test just to make the failure easier for users to find in the long string of text from all of the unit tests.